infer/function.go

```diff
@@ -29,11 +29,23 @@ import (
 	"github.com/pulumi/pulumi-go-provider/middleware/schema"
 )
 
+// FunctionRequest wraps the input type for a function call
+type FunctionRequest[I any] struct {
+	// Input contains the function input arguments.
+	Input I
+}
+
+// FunctionResponse wraps the output type for a function call
+type FunctionResponse[O any] struct {
+	// Output contains the function result.
+	Output O
+}
+
 // Fn is a function (also called fnvoke) inferred from code. `I` is the function input,
 // and `O` is the function output. Both must be structs.
 type Fn[I any, O any] interface {
 	// A function is a mapping from `I` to `O`.
-	Call(ctx context.Context, input I) (output O, err error)
+	Call(ctx context.Context, req FunctionRequest[I]) (resp FunctionResponse[O], err error)
```

Comment:
I propose we rename this to `Invoke` to greatly reduce confusion vis-a-vis resource methods.

Reply:
I've split the `Call` → `Invoke` rename into a separate PR ([#329](https://github.com/pulumi/pulumi-go-provider/pull/329)) to keep this PR focused on the request/response struct refactor. This should make review easier and help isolate the breaking changes.